ftw/file/browser/upload.py

```diff
@@ -35,3 +50,113 @@ def __call__(self):
+        if not IFolderish.providedBy(context):
+            context = aq_parent(context)
+
+        request = context.REQUEST
+        ctr_tool = getToolByName(context, 'content_type_registry')
+        utility = getToolByName(context, 'portal_tinymce')
+
+        id = request['uploadfile'].filename
+        content_type = request['uploadfile'].headers["Content-Type"]
+        typename = ctr_tool.findTypeName(id, content_type, "")
+
+        # check mime type to make sure an image is uploaded
+        if not is_image(content_type):
+            return self.errorMessage(
+                translate(_('Only image upload allowed.'),
+                          context=self.request))
+
+        # Permission checks based on code by Danny Bloemendaal
+
+        # 1) check if the current user has permissions to add stuff
+        if not context.portal_membership.checkPermission(
+            'Add portal content', context):
+            return self.errorMessage(
+                "You do not have permission to upload files in this folder")
+
+        # 2) check image types uploadable in folder.
+        #    priority is to content_type_registry image type
+        allowed_types = [t.id for t in context.getAllowedTypes()]
+        if typename in allowed_types:
+            uploadable_types = [typename]
+        else:
+            uploadable_types = []
+
+        if content_type.split('/')[0] == 'image':
+            image_portal_types = utility.imageobjects.split('\n')
+            uploadable_types += [t for t in image_portal_types
+                                    if t in allowed_types
+                                       and t not in uploadable_types]
+
+        # Get an unused filename without path
+        id = self.cleanupFilename(id)
+
+        for metatype in uploadable_types:
+            try:
+                newid = context.invokeFactory(type_name=metatype, id=id)
+                if newid is None or newid == '':
+                    newid = id
+                break
+            except ValueError:
+                continue
+            except BadRequest:
+                return self.errorMessage(
+                    translate(_("Bad filename, please rename."),
+                              context=self.request))
+        else:
+            return self.errorMessage(translate(
+                _("Not allowed to upload a file of this type to this folder"),
+                context=self.request))
+
+        obj = getattr(context, newid, None)
+
+        # Set title + description.
+        # Attempt to use Archetypes mutator if there is one, in case it uses
+        # a custom storage
+        title = request['uploadtitle']
+        description = request['uploaddescription']
+
+        if description:
+            try:
+                obj.setDescription(description)
+            except AttributeError:
+                obj.description = description
+
+        if HAS_DEXTERITY and IDexterityContent.providedBy(obj):
```

Comment:
ah, sure, the other adapter is specific..
using ``IBaseFolder`` is okay, I'd try to avoid using ``overrides.zcml`` here, because overriding overrides is quite hard.. :wink: